{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "455a77eb-42ae-4f22-bb0a-831152588c6e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## CHAID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed9d7a1-c359-47f3-beb9-e995e76511bb",
   "metadata": {},
   "source": [
    "Chi-square Automatic Interaction Detector (CHAID) foi uma técnica criada por Gordon V. Kass em 1980.  CHAID é uma ferramenta utilizada para descobrir a relação entre as variáveis.  A análise CHAID constrói um modelo preditivo, ou árvore, para ajudar a determinar como as variáveis melhor se fundem para explicar o resultado na variável dependente dada. Na análise CHAID, dados nominais, ordinais e contínuos podem ser usados, onde os preditores contínuos são divididos em categorias com aproximadamente igual número de observações.\n",
    "\n",
    "Fonte: estatistica.pt/chi-square-automatic-interaction-detector-chaid/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35c40bcf-cb35-4ebd-b31f-ee6427e218bb",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Exemplo 1: biblioteca CHAID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f11fddf7-b558-42c0-b5bb-0dae34af4c88",
   "metadata": {},
   "source": [
    "Doc: https://github.com/Rambatino/CHAID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "106d70a3-73c9-436b-96ca-0cd588363a95",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: CHAID in /opt/conda/lib/python3.9/site-packages (5.3.0)\n",
      "Requirement already satisfied: colorlover in /opt/conda/lib/python3.9/site-packages (from CHAID) (0.3.0)\n",
      "Requirement already satisfied: numpy in /opt/conda/lib/python3.9/site-packages (from CHAID) (1.21.6)\n",
      "Requirement already satisfied: cython in /opt/conda/lib/python3.9/site-packages (from CHAID) (0.29.28)\n",
      "Requirement already satisfied: treelib in /opt/conda/lib/python3.9/site-packages (from CHAID) (1.6.1)\n",
      "Requirement already satisfied: graphviz in /opt/conda/lib/python3.9/site-packages (from CHAID) (0.20.1)\n",
      "Requirement already satisfied: pandas in /opt/conda/lib/python3.9/site-packages (from CHAID) (1.4.2)\n",
      "Requirement already satisfied: plotly in /opt/conda/lib/python3.9/site-packages (from CHAID) (5.10.0)\n",
      "Requirement already satisfied: pytest in /opt/conda/lib/python3.9/site-packages (from CHAID) (7.1.2)\n",
      "Requirement already satisfied: scipy in /opt/conda/lib/python3.9/site-packages (from CHAID) (1.8.0)\n",
      "Requirement already satisfied: savReaderWriter in /opt/conda/lib/python3.9/site-packages (from CHAID) (3.4.2)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /opt/conda/lib/python3.9/site-packages (from pandas->CHAID) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /opt/conda/lib/python3.9/site-packages (from pandas->CHAID) (2022.1)\n",
      "Requirement already satisfied: tenacity>=6.2.0 in /opt/conda/lib/python3.9/site-packages (from plotly->CHAID) (8.0.1)\n",
      "Requirement already satisfied: py>=1.8.2 in /opt/conda/lib/python3.9/site-packages (from pytest->CHAID) (1.11.0)\n",
      "Requirement already satisfied: iniconfig in /opt/conda/lib/python3.9/site-packages (from pytest->CHAID) (1.1.1)\n",
      "Requirement already satisfied: tomli>=1.0.0 in /opt/conda/lib/python3.9/site-packages (from pytest->CHAID) (2.0.1)\n",
      "Requirement already satisfied: attrs>=19.2.0 in /opt/conda/lib/python3.9/site-packages (from pytest->CHAID) (21.4.0)\n",
      "Requirement already satisfied: pluggy<2.0,>=0.12 in /opt/conda/lib/python3.9/site-packages (from pytest->CHAID) (1.0.0)\n",
      "Requirement already satisfied: packaging in /opt/conda/lib/python3.9/site-packages (from pytest->CHAID) (21.3)\n",
      "Requirement already satisfied: future in /opt/conda/lib/python3.9/site-packages (from treelib->CHAID) (0.18.2)\n",
      "Requirement already satisfied: six>=1.5 in /opt/conda/lib/python3.9/site-packages (from python-dateutil>=2.8.1->pandas->CHAID) (1.16.0)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /opt/conda/lib/python3.9/site-packages (from packaging->pytest->CHAID) (3.0.9)\n"
     ]
    }
   ],
   "source": [
    "!pip install CHAID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8f3790fc-7461-4202-b1de-457fa045aab4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from CHAID import Tree, NominalColumn\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d129f255-eedb-473d-b462-6158833a1be7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "      <th>c</th>\n",
       "      <th>d</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>379.011672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>271.746426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>265.633091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>336.329989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>219.477118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>344.723591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>235.304306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>352.331289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>352.804308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>208.421965</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   a  b  c           d\n",
       "0  1  2  3  379.011672\n",
       "1  1  2  3  271.746426\n",
       "2  1  2  3  265.633091\n",
       "3  1  2  3  336.329989\n",
       "4  1  2  3  219.477118\n",
       "5  2  2  3  344.723591\n",
       "6  2  2  3  235.304306\n",
       "7  2  2  3  352.331289\n",
       "8  2  2  3  352.804308\n",
       "9  2  2  3  208.421965"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Criação dos dados \n",
    "ndarr = np.array(([1, 2, 3] * 5) + ([2, 2, 3] * 5)).reshape(10, 3)\n",
    "df = pd.DataFrame(ndarr)\n",
    "df.columns = ['a', 'b', 'c']\n",
    "arr = np.array(([1] * 5) + ([2] * 5))\n",
    "df['d'] = np.random.normal(300, 100, 10)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cd1e474f-9b8d-41c7-b989-7a15ecb1737b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "([], {1: 5.0, 2: 5.0}, (a, p=0.001565402258002549, score=10.0, groups=[[1], [2]]), dof=1))\n",
      "|-- ([1], {1: 5.0, 2: 0}, <Invalid Chaid Split> - the node only contains single category respondents)\n",
      "+-- ([2], {1: 0, 2: 5.0}, <Invalid Chaid Split> - the node only contains single category respondents)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Estabelece os parâmetros de entrada\n",
    "independent_variable_columns = ['a', 'b', 'c']\n",
    "dep_variable = 'd'\n",
    "\n",
    "# Cria a árvore\n",
    "cols = [\n",
    "  NominalColumn(ndarr[:,0], name='a'),\n",
    "  NominalColumn(ndarr[:,1], name='b'),\n",
    "  NominalColumn(ndarr[:,2], name='c')\n",
    "]\n",
    "tree = Tree(cols, NominalColumn(arr, name='d'), {'min_child_node_size': 1, 'min_parent_node_size':1, 'dep_variable_type':'continuous'})\n",
    "z\n",
    "tree.print_tree()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f0296ae6-b4b8-4d8c-a3b1-7c7e936c2d35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<treelib.tree.Tree at 0x7fb730f6fbb0>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Para extrair um objeto do tipo LibTree\n",
    "tree.to_tree()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b48867df-a405-4ca2-a4ee-6fa3a796beb3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting kaleido\n",
      "  Using cached kaleido-0.2.1-py2.py3-none-manylinux1_x86_64.whl (79.9 MB)\n",
      "Installing collected packages: kaleido\n",
      "Successfully installed kaleido-0.2.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install -U kaleido"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2db4382c-e3aa-47cd-9cca-51ca4eb61793",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Transform failed with error code 525: Failed to execute 'getPointAtLength' on 'SVGGeometryElement': The element's path is empty.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [7]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mtree\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrender\u001b[49m\u001b[43m(\u001b[49m\u001b[43mview\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m.\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/CHAID/tree.py:291\u001b[0m, in \u001b[0;36mTree.render\u001b[0;34m(self, path, view)\u001b[0m\n\u001b[1;32m    290\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrender\u001b[39m(\u001b[38;5;28mself\u001b[39m, path\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, view\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m):\n\u001b[0;32m--> 291\u001b[0m     \u001b[43mGraph\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrender\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpath\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mview\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/CHAID/graph.py:72\u001b[0m, in \u001b[0;36mGraph.render\u001b[0;34m(self, path, view)\u001b[0m\n\u001b[1;32m     66\u001b[0m g \u001b[38;5;241m=\u001b[39m Digraph(\n\u001b[1;32m     67\u001b[0m     \u001b[38;5;28mformat\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpng\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     68\u001b[0m     graph_attr\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msplines\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mortho\u001b[39m\u001b[38;5;124m\"\u001b[39m},\n\u001b[1;32m     69\u001b[0m     node_attr\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mshape\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplaintext\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlabelloc\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m},\n\u001b[1;32m     70\u001b[0m )\n\u001b[1;32m     71\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m node \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtree:\n\u001b[0;32m---> 72\u001b[0m     image \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbar_chart\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnode\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     73\u001b[0m     g\u001b[38;5;241m.\u001b[39mnode(\u001b[38;5;28mstr\u001b[39m(node\u001b[38;5;241m.\u001b[39mnode_id), image\u001b[38;5;241m=\u001b[39mimage)\n\u001b[1;32m     74\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m node\u001b[38;5;241m.\u001b[39mparent \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/CHAID/graph.py:96\u001b[0m, in \u001b[0;36mGraph.bar_chart\u001b[0;34m(self, node)\u001b[0m\n\u001b[1;32m     93\u001b[0m     fig[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdata\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39mappend(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_table(node))\n\u001b[1;32m     95\u001b[0m filename \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtempdir, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnode-\u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m.png\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(node\u001b[38;5;241m.\u001b[39mnode_id))\n\u001b[0;32m---> 96\u001b[0m \u001b[43mpio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrite_image\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfig\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfile\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mformat\u001b[39;49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpng\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     97\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m filename\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/plotly/io/_kaleido.py:267\u001b[0m, in \u001b[0;36mwrite_image\u001b[0;34m(fig, file, format, scale, width, height, validate, engine)\u001b[0m\n\u001b[1;32m    251\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    252\u001b[0m                 \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    253\u001b[0m \u001b[38;5;124;03mCannot infer image type from output path '{file}'.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    261\u001b[0m                 )\n\u001b[1;32m    262\u001b[0m             )\n\u001b[1;32m    264\u001b[0m     \u001b[38;5;66;03m# Request image\u001b[39;00m\n\u001b[1;32m    265\u001b[0m     \u001b[38;5;66;03m# -------------\u001b[39;00m\n\u001b[1;32m    266\u001b[0m     \u001b[38;5;66;03m# Do this first so we don't create a file if image conversion fails\u001b[39;00m\n\u001b[0;32m--> 267\u001b[0m     img_data \u001b[38;5;241m=\u001b[39m \u001b[43mto_image\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    268\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfig\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    269\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mformat\u001b[39;49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mformat\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    270\u001b[0m \u001b[43m        \u001b[49m\u001b[43mscale\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mscale\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    271\u001b[0m \u001b[43m        \u001b[49m\u001b[43mwidth\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mwidth\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    272\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheight\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheight\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    273\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalidate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mvalidate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    274\u001b[0m \u001b[43m        \u001b[49m\u001b[43mengine\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    275\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    277\u001b[0m     \u001b[38;5;66;03m# Open file\u001b[39;00m\n\u001b[1;32m    278\u001b[0m     \u001b[38;5;66;03m# ---------\u001b[39;00m\n\u001b[1;32m    279\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m path \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    280\u001b[0m         \u001b[38;5;66;03m# We previously failed to make sense of `file` as a pathlib object.\u001b[39;00m\n\u001b[1;32m    281\u001b[0m         \u001b[38;5;66;03m# Attempt to write to `file` as an open file descriptor.\u001b[39;00m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/plotly/io/_kaleido.py:144\u001b[0m, in \u001b[0;36mto_image\u001b[0;34m(fig, format, width, height, scale, validate, engine)\u001b[0m\n\u001b[1;32m    141\u001b[0m \u001b[38;5;66;03m# Validate figure\u001b[39;00m\n\u001b[1;32m    142\u001b[0m \u001b[38;5;66;03m# ---------------\u001b[39;00m\n\u001b[1;32m    143\u001b[0m fig_dict \u001b[38;5;241m=\u001b[39m validate_coerce_fig_to_dict(fig, validate)\n\u001b[0;32m--> 144\u001b[0m img_bytes \u001b[38;5;241m=\u001b[39m \u001b[43mscope\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtransform\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    145\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfig_dict\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mformat\u001b[39;49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mformat\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwidth\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mwidth\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheight\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheight\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mscale\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mscale\u001b[49m\n\u001b[1;32m    146\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    148\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m img_bytes\n",
      "File \u001b[0;32m~/resources/venv/lib/python3.9/site-packages/kaleido/scopes/plotly.py:161\u001b[0m, in \u001b[0;36mPlotlyScope.transform\u001b[0;34m(self, figure, format, width, height, scale)\u001b[0m\n\u001b[1;32m    159\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m code \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    160\u001b[0m     message \u001b[38;5;241m=\u001b[39m response\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmessage\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[0;32m--> 161\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    162\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTransform failed with error code \u001b[39m\u001b[38;5;132;01m{code}\u001b[39;00m\u001b[38;5;124m: \u001b[39m\u001b[38;5;132;01m{message}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(\n\u001b[1;32m    163\u001b[0m             code\u001b[38;5;241m=\u001b[39mcode, message\u001b[38;5;241m=\u001b[39mmessage\n\u001b[1;32m    164\u001b[0m         )\n\u001b[1;32m    165\u001b[0m     )\n\u001b[1;32m    167\u001b[0m img \u001b[38;5;241m=\u001b[39m response\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresult\u001b[39m\u001b[38;5;124m\"\u001b[39m)\u001b[38;5;241m.\u001b[39mencode(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mutf-8\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    169\u001b[0m \u001b[38;5;66;03m# Base64 decode binary types\u001b[39;00m\n",
      "\u001b[0;31mValueError\u001b[0m: Transform failed with error code 525: Failed to execute 'getPointAtLength' on 'SVGGeometryElement': The element's path is empty."
     ]
    }
   ],
   "source": [
    "tree.render(view=True, path='.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "af77f11a-fb08-4e8a-83aa-d5bc75cc7cca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([], {1: 5.0, 2: 5.0}, (a, p=0.001565402258002549, score=10.0, groups=[[1], [2]]), dof=1))"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Os diferentes nós da árvore podem ser acessados assim:\n",
    "first_node = tree.tree_store[0]\n",
    "\n",
    "first_node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c1262205-ab2a-421c-87c3-ec2d533e5e04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: 5.0, 2: 5.0}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Os parâmetros dos nós pondem ser acessados assim\n",
    "first_node.members"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cdf006e2-1d00-47e8-b5b2-d96b7094d841",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.001565402258002549\n",
      "10.0\n"
     ]
    }
   ],
   "source": [
    "# As propriedades podem ser acessadas assim\n",
    "print(first_node.split.p)\n",
    "print(first_node.split.score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "eda4f0d3-68aa-436b-aa45-4ce0c0f8464c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cloning into 'py-earth'...\n",
      "fatal: unable to connect to github.com:\n",
      "github.com[0: 140.82.112.3]: errno=Connection timed out\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!git clone git://github.com/scikit-learn-contrib/py-earth.git\n",
    "# cd py-earth\n",
    "# sudo python setup.py install"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cea56ec-5186-4bfd-bb3f-5b46bdbab2e2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting sklearn-contrib-py-earth\n",
      "  Downloading sklearn-contrib-py-earth-0.1.0.tar.gz (1.0 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.0/1.0 MB\u001b[0m \u001b[31m55.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: scipy>=0.16 in /opt/conda/lib/python3.9/site-packages (from sklearn-contrib-py-earth) (1.8.0)\n",
      "Requirement already satisfied: scikit-learn>=0.16 in /opt/conda/lib/python3.9/site-packages (from sklearn-contrib-py-earth) (1.1.0)\n",
      "Requirement already satisfied: six in /opt/conda/lib/python3.9/site-packages (from sklearn-contrib-py-earth) (1.16.0)\n",
      "Requirement already satisfied: numpy>=1.17.3 in /opt/conda/lib/python3.9/site-packages (from scikit-learn>=0.16->sklearn-contrib-py-earth) (1.21.6)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /opt/conda/lib/python3.9/site-packages (from scikit-learn>=0.16->sklearn-contrib-py-earth) (3.1.0)\n",
      "Requirement already satisfied: joblib>=1.0.0 in /opt/conda/lib/python3.9/site-packages (from scikit-learn>=0.16->sklearn-contrib-py-earth) (1.1.0)\n",
      "Building wheels for collected packages: sklearn-contrib-py-earth\n",
      "  Building wheel for sklearn-contrib-py-earth (setup.py) ... \u001b[?25lerror\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mpython setup.py bdist_wheel\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[254 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m /opt/conda/lib/python3.9/site-packages/setuptools/dist.py:767: UserWarning: Usage of dash-separated 'description-file' will not be supported in future versions. Please use the underscore name 'description_file' instead\n",
      "  \u001b[31m   \u001b[0m   warnings.warn(\n",
      "  \u001b[31m   \u001b[0m running bdist_wheel\n",
      "  \u001b[31m   \u001b[0m running build\n",
      "  \u001b[31m   \u001b[0m running build_py\n",
      "  \u001b[31m   \u001b[0m creating build\n",
      "  \u001b[31m   \u001b[0m creating build/lib.linux-x86_64-cpython-39\n",
      "  \u001b[31m   \u001b[0m creating build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_version.py -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/export.py -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/earth.py -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/__init__.py -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m creating build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/test_knot_search.py -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/test_util.py -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/test_qr.py -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/test_earth.py -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/test_forward.py -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/testing_utils.py -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/test_export.py -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/test_pruning.py -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/__init__.py -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m creating build/lib.linux-x86_64-cpython-39/pyearth/test/basis\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/basis/test_constant.py -> build/lib.linux-x86_64-cpython-39/pyearth/test/basis\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/basis/test_linear.py -> build/lib.linux-x86_64-cpython-39/pyearth/test/basis\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/basis/test_smoothed_hinge.py -> build/lib.linux-x86_64-cpython-39/pyearth/test/basis\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/basis/test_missingness.py -> build/lib.linux-x86_64-cpython-39/pyearth/test/basis\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/basis/base.py -> build/lib.linux-x86_64-cpython-39/pyearth/test/basis\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/basis/test_basis.py -> build/lib.linux-x86_64-cpython-39/pyearth/test/basis\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/basis/test_hinge.py -> build/lib.linux-x86_64-cpython-39/pyearth/test/basis\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/basis/__init__.py -> build/lib.linux-x86_64-cpython-39/pyearth/test/basis\n",
      "  \u001b[31m   \u001b[0m creating build/lib.linux-x86_64-cpython-39/pyearth/test/record\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/record/test_forward_pass.py -> build/lib.linux-x86_64-cpython-39/pyearth/test/record\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/record/test_pruning_pass.py -> build/lib.linux-x86_64-cpython-39/pyearth/test/record\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/record/__init__.py -> build/lib.linux-x86_64-cpython-39/pyearth/test/record\n",
      "  \u001b[31m   \u001b[0m running egg_info\n",
      "  \u001b[31m   \u001b[0m writing sklearn_contrib_py_earth.egg-info/PKG-INFO\n",
      "  \u001b[31m   \u001b[0m writing dependency_links to sklearn_contrib_py_earth.egg-info/dependency_links.txt\n",
      "  \u001b[31m   \u001b[0m writing requirements to sklearn_contrib_py_earth.egg-info/requires.txt\n",
      "  \u001b[31m   \u001b[0m writing top-level names to sklearn_contrib_py_earth.egg-info/top_level.txt\n",
      "  \u001b[31m   \u001b[0m reading manifest file 'sklearn_contrib_py_earth.egg-info/SOURCES.txt'\n",
      "  \u001b[31m   \u001b[0m reading manifest template 'MANIFEST.in'\n",
      "  \u001b[31m   \u001b[0m warning: no files found matching 'pyearth/test/pathological_data'\n",
      "  \u001b[31m   \u001b[0m adding license file 'LICENSE.txt'\n",
      "  \u001b[31m   \u001b[0m writing manifest file 'sklearn_contrib_py_earth.egg-info/SOURCES.txt'\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_basis.c -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_basis.pxd -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_forward.c -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_forward.pxd -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_knot_search.c -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_knot_search.pxd -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_pruning.c -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_pruning.pxd -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_qr.c -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_qr.pxd -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_record.c -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_record.pxd -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_types.c -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_types.pxd -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_util.c -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/_util.pxd -> build/lib.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/earth_linvars_regress.txt -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/earth_regress.txt -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/earth_regress_missing_data.txt -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/earth_regress_smooth.txt -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/forward_regress.txt -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m copying pyearth/test/test_data.csv -> build/lib.linux-x86_64-cpython-39/pyearth/test\n",
      "  \u001b[31m   \u001b[0m UPDATING build/lib.linux-x86_64-cpython-39/pyearth/_version.py\n",
      "  \u001b[31m   \u001b[0m set build/lib.linux-x86_64-cpython-39/pyearth/_version.py to '0.1.0'\n",
      "  \u001b[31m   \u001b[0m running build_ext\n",
      "  \u001b[31m   \u001b[0m building 'pyearth._util' extension\n",
      "  \u001b[31m   \u001b[0m creating build/temp.linux-x86_64-cpython-39\n",
      "  \u001b[31m   \u001b[0m creating build/temp.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m gcc -pthread -B /opt/conda/compiler_compat -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /opt/conda/include -fPIC -O2 -isystem /opt/conda/include -fPIC -I/opt/conda/lib/python3.9/site-packages/numpy/core/include -I/opt/conda/include/python3.9 -c pyearth/_util.c -o build/temp.linux-x86_64-cpython-39/pyearth/_util.o\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/lib/python3.9/site-packages/numpy/core/include/numpy/ndarraytypes.h:1969,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/lib/python3.9/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/lib/python3.9/site-packages/numpy/core/include/numpy/arrayobject.h:4,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:495:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/lib/python3.9/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning \"Using deprecated NumPy API, disable it with \" \"#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION\" [-Wcpp]\n",
      "  \u001b[31m   \u001b[0m    17 | #warning \"Using deprecated NumPy API, disable it with \" \\\n",
      "  \u001b[31m   \u001b[0m       |  ^~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c: In function ‘__Pyx_ParseOptionalKeywords’:\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7561:21: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7561 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                     ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7561:21: warning: ‘PyUnicode_AsUnicode’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7561 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                     ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here\n",
      "  \u001b[31m   \u001b[0m   580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(\n",
      "  \u001b[31m   \u001b[0m       |                                             ^~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7561:21: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7561 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                     ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7561:21: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7561 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                     ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7561:21: warning: ‘PyUnicode_AsUnicode’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7561 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                     ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here\n",
      "  \u001b[31m   \u001b[0m   580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(\n",
      "  \u001b[31m   \u001b[0m       |                                             ^~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7561:21: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7561 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                     ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7577:25: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7577 |                         (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                         ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7577:25: warning: ‘PyUnicode_AsUnicode’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7577 |                         (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                         ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here\n",
      "  \u001b[31m   \u001b[0m   580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(\n",
      "  \u001b[31m   \u001b[0m       |                                             ^~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7577:25: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7577 |                         (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                         ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7577:25: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7577 |                         (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                         ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7577:25: warning: ‘PyUnicode_AsUnicode’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7577 |                         (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                         ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here\n",
      "  \u001b[31m   \u001b[0m   580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(\n",
      "  \u001b[31m   \u001b[0m       |                                             ^~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7577:25: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7577 |                         (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                         ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c: In function ‘__Pyx__ExceptionSave’:\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8071:21: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?\n",
      "  \u001b[31m   \u001b[0m  8071 |     *type = tstate->exc_type;\n",
      "  \u001b[31m   \u001b[0m       |                     ^~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                     curexc_type\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8072:22: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?\n",
      "  \u001b[31m   \u001b[0m  8072 |     *value = tstate->exc_value;\n",
      "  \u001b[31m   \u001b[0m       |                      ^~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                      curexc_value\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8073:19: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?\n",
      "  \u001b[31m   \u001b[0m  8073 |     *tb = tstate->exc_traceback;\n",
      "  \u001b[31m   \u001b[0m       |                   ^~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                   curexc_traceback\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c: In function ‘__Pyx__ExceptionReset’:\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8080:24: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?\n",
      "  \u001b[31m   \u001b[0m  8080 |     tmp_type = tstate->exc_type;\n",
      "  \u001b[31m   \u001b[0m       |                        ^~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                        curexc_type\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8081:25: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?\n",
      "  \u001b[31m   \u001b[0m  8081 |     tmp_value = tstate->exc_value;\n",
      "  \u001b[31m   \u001b[0m       |                         ^~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                         curexc_value\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8082:22: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?\n",
      "  \u001b[31m   \u001b[0m  8082 |     tmp_tb = tstate->exc_traceback;\n",
      "  \u001b[31m   \u001b[0m       |                      ^~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                      curexc_traceback\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8083:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?\n",
      "  \u001b[31m   \u001b[0m  8083 |     tstate->exc_type = type;\n",
      "  \u001b[31m   \u001b[0m       |             ^~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |             curexc_type\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8084:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?\n",
      "  \u001b[31m   \u001b[0m  8084 |     tstate->exc_value = value;\n",
      "  \u001b[31m   \u001b[0m       |             ^~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |             curexc_value\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8085:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?\n",
      "  \u001b[31m   \u001b[0m  8085 |     tstate->exc_traceback = tb;\n",
      "  \u001b[31m   \u001b[0m       |             ^~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |             curexc_traceback\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c: In function ‘__Pyx__GetException’:\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8140:24: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?\n",
      "  \u001b[31m   \u001b[0m  8140 |     tmp_type = tstate->exc_type;\n",
      "  \u001b[31m   \u001b[0m       |                        ^~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                        curexc_type\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8141:25: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?\n",
      "  \u001b[31m   \u001b[0m  8141 |     tmp_value = tstate->exc_value;\n",
      "  \u001b[31m   \u001b[0m       |                         ^~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                         curexc_value\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8142:22: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?\n",
      "  \u001b[31m   \u001b[0m  8142 |     tmp_tb = tstate->exc_traceback;\n",
      "  \u001b[31m   \u001b[0m       |                      ^~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                      curexc_traceback\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8143:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?\n",
      "  \u001b[31m   \u001b[0m  8143 |     tstate->exc_type = local_type;\n",
      "  \u001b[31m   \u001b[0m       |             ^~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |             curexc_type\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8144:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?\n",
      "  \u001b[31m   \u001b[0m  8144 |     tstate->exc_value = local_value;\n",
      "  \u001b[31m   \u001b[0m       |             ^~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |             curexc_value\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8145:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?\n",
      "  \u001b[31m   \u001b[0m  8145 |     tstate->exc_traceback = local_tb;\n",
      "  \u001b[31m   \u001b[0m       |             ^~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |             curexc_traceback\n",
      "  \u001b[31m   \u001b[0m error: command '/usr/bin/gcc' failed with exit code 1\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[31m  ERROR: Failed building wheel for sklearn-contrib-py-earth\u001b[0m\u001b[31m\n",
      "\u001b[0m\u001b[?25h  Running setup.py clean for sklearn-contrib-py-earth\n",
      "Failed to build sklearn-contrib-py-earth\n",
      "Installing collected packages: sklearn-contrib-py-earth\n",
      "  Running setup.py install for sklearn-contrib-py-earth ... \u001b[?25lerror\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mRunning setup.py install for sklearn-contrib-py-earth\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[204 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m /opt/conda/lib/python3.9/site-packages/setuptools/dist.py:767: UserWarning: Usage of dash-separated 'description-file' will not be supported in future versions. Please use the underscore name 'description_file' instead\n",
      "  \u001b[31m   \u001b[0m   warnings.warn(\n",
      "  \u001b[31m   \u001b[0m running install\n",
      "  \u001b[31m   \u001b[0m /opt/conda/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.\n",
      "  \u001b[31m   \u001b[0m   warnings.warn(\n",
      "  \u001b[31m   \u001b[0m running build\n",
      "  \u001b[31m   \u001b[0m running build_py\n",
      "  \u001b[31m   \u001b[0m running egg_info\n",
      "  \u001b[31m   \u001b[0m writing sklearn_contrib_py_earth.egg-info/PKG-INFO\n",
      "  \u001b[31m   \u001b[0m writing dependency_links to sklearn_contrib_py_earth.egg-info/dependency_links.txt\n",
      "  \u001b[31m   \u001b[0m writing requirements to sklearn_contrib_py_earth.egg-info/requires.txt\n",
      "  \u001b[31m   \u001b[0m writing top-level names to sklearn_contrib_py_earth.egg-info/top_level.txt\n",
      "  \u001b[31m   \u001b[0m reading manifest file 'sklearn_contrib_py_earth.egg-info/SOURCES.txt'\n",
      "  \u001b[31m   \u001b[0m reading manifest template 'MANIFEST.in'\n",
      "  \u001b[31m   \u001b[0m warning: no files found matching 'pyearth/test/pathological_data'\n",
      "  \u001b[31m   \u001b[0m adding license file 'LICENSE.txt'\n",
      "  \u001b[31m   \u001b[0m writing manifest file 'sklearn_contrib_py_earth.egg-info/SOURCES.txt'\n",
      "  \u001b[31m   \u001b[0m UPDATING build/lib.linux-x86_64-cpython-39/pyearth/_version.py\n",
      "  \u001b[31m   \u001b[0m set build/lib.linux-x86_64-cpython-39/pyearth/_version.py to '0.1.0'\n",
      "  \u001b[31m   \u001b[0m running build_ext\n",
      "  \u001b[31m   \u001b[0m building 'pyearth._util' extension\n",
      "  \u001b[31m   \u001b[0m creating build/temp.linux-x86_64-cpython-39\n",
      "  \u001b[31m   \u001b[0m creating build/temp.linux-x86_64-cpython-39/pyearth\n",
      "  \u001b[31m   \u001b[0m gcc -pthread -B /opt/conda/compiler_compat -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /opt/conda/include -fPIC -O2 -isystem /opt/conda/include -fPIC -I/opt/conda/lib/python3.9/site-packages/numpy/core/include -I/opt/conda/include/python3.9 -c pyearth/_util.c -o build/temp.linux-x86_64-cpython-39/pyearth/_util.o\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/lib/python3.9/site-packages/numpy/core/include/numpy/ndarraytypes.h:1969,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/lib/python3.9/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/lib/python3.9/site-packages/numpy/core/include/numpy/arrayobject.h:4,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:495:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/lib/python3.9/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning \"Using deprecated NumPy API, disable it with \" \"#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION\" [-Wcpp]\n",
      "  \u001b[31m   \u001b[0m    17 | #warning \"Using deprecated NumPy API, disable it with \" \\\n",
      "  \u001b[31m   \u001b[0m       |  ^~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c: In function ‘__Pyx_ParseOptionalKeywords’:\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7561:21: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7561 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                     ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7561:21: warning: ‘PyUnicode_AsUnicode’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7561 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                     ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here\n",
      "  \u001b[31m   \u001b[0m   580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(\n",
      "  \u001b[31m   \u001b[0m       |                                             ^~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7561:21: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7561 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                     ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7561:21: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7561 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                     ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7561:21: warning: ‘PyUnicode_AsUnicode’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7561 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                     ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here\n",
      "  \u001b[31m   \u001b[0m   580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(\n",
      "  \u001b[31m   \u001b[0m       |                                             ^~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7561:21: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7561 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                     ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7577:25: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7577 |                         (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                         ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7577:25: warning: ‘PyUnicode_AsUnicode’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7577 |                         (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                         ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here\n",
      "  \u001b[31m   \u001b[0m   580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(\n",
      "  \u001b[31m   \u001b[0m       |                                             ^~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7577:25: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7577 |                         (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                         ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7577:25: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7577 |                         (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                         ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7577:25: warning: ‘PyUnicode_AsUnicode’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7577 |                         (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                         ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here\n",
      "  \u001b[31m   \u001b[0m   580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(\n",
      "  \u001b[31m   \u001b[0m       |                                             ^~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:7577:25: warning: ‘_PyUnicode_get_wstr_length’ is deprecated [-Wdeprecated-declarations]\n",
      "  \u001b[31m   \u001b[0m  7577 |                         (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :\n",
      "  \u001b[31m   \u001b[0m       |                         ^\n",
      "  \u001b[31m   \u001b[0m In file included from /opt/conda/include/python3.9/unicodeobject.h:1026,\n",
      "  \u001b[31m   \u001b[0m                  from /opt/conda/include/python3.9/Python.h:93,\n",
      "  \u001b[31m   \u001b[0m                  from pyearth/_util.c:24:\n",
      "  \u001b[31m   \u001b[0m /opt/conda/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here\n",
      "  \u001b[31m   \u001b[0m   446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {\n",
      "  \u001b[31m   \u001b[0m       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c: In function ‘__Pyx__ExceptionSave’:\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8071:21: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?\n",
      "  \u001b[31m   \u001b[0m  8071 |     *type = tstate->exc_type;\n",
      "  \u001b[31m   \u001b[0m       |                     ^~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                     curexc_type\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8072:22: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?\n",
      "  \u001b[31m   \u001b[0m  8072 |     *value = tstate->exc_value;\n",
      "  \u001b[31m   \u001b[0m       |                      ^~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                      curexc_value\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8073:19: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?\n",
      "  \u001b[31m   \u001b[0m  8073 |     *tb = tstate->exc_traceback;\n",
      "  \u001b[31m   \u001b[0m       |                   ^~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                   curexc_traceback\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c: In function ‘__Pyx__ExceptionReset’:\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8080:24: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?\n",
      "  \u001b[31m   \u001b[0m  8080 |     tmp_type = tstate->exc_type;\n",
      "  \u001b[31m   \u001b[0m       |                        ^~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                        curexc_type\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8081:25: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?\n",
      "  \u001b[31m   \u001b[0m  8081 |     tmp_value = tstate->exc_value;\n",
      "  \u001b[31m   \u001b[0m       |                         ^~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                         curexc_value\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8082:22: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?\n",
      "  \u001b[31m   \u001b[0m  8082 |     tmp_tb = tstate->exc_traceback;\n",
      "  \u001b[31m   \u001b[0m       |                      ^~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                      curexc_traceback\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8083:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?\n",
      "  \u001b[31m   \u001b[0m  8083 |     tstate->exc_type = type;\n",
      "  \u001b[31m   \u001b[0m       |             ^~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |             curexc_type\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8084:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?\n",
      "  \u001b[31m   \u001b[0m  8084 |     tstate->exc_value = value;\n",
      "  \u001b[31m   \u001b[0m       |             ^~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |             curexc_value\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8085:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?\n",
      "  \u001b[31m   \u001b[0m  8085 |     tstate->exc_traceback = tb;\n",
      "  \u001b[31m   \u001b[0m       |             ^~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |             curexc_traceback\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c: In function ‘__Pyx__GetException’:\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8140:24: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?\n",
      "  \u001b[31m   \u001b[0m  8140 |     tmp_type = tstate->exc_type;\n",
      "  \u001b[31m   \u001b[0m       |                        ^~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                        curexc_type\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8141:25: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?\n",
      "  \u001b[31m   \u001b[0m  8141 |     tmp_value = tstate->exc_value;\n",
      "  \u001b[31m   \u001b[0m       |                         ^~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                         curexc_value\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8142:22: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?\n",
      "  \u001b[31m   \u001b[0m  8142 |     tmp_tb = tstate->exc_traceback;\n",
      "  \u001b[31m   \u001b[0m       |                      ^~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |                      curexc_traceback\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8143:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?\n",
      "  \u001b[31m   \u001b[0m  8143 |     tstate->exc_type = local_type;\n",
      "  \u001b[31m   \u001b[0m       |             ^~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |             curexc_type\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8144:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?\n",
      "  \u001b[31m   \u001b[0m  8144 |     tstate->exc_value = local_value;\n",
      "  \u001b[31m   \u001b[0m       |             ^~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |             curexc_value\n",
      "  \u001b[31m   \u001b[0m pyearth/_util.c:8145:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?\n",
      "  \u001b[31m   \u001b[0m  8145 |     tstate->exc_traceback = local_tb;\n",
      "  \u001b[31m   \u001b[0m       |             ^~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m       |             curexc_traceback\n",
      "  \u001b[31m   \u001b[0m error: command '/usr/bin/gcc' failed with exit code 1\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[1;31merror\u001b[0m: \u001b[1mlegacy-install-failure\u001b[0m\n",
      "\n",
      "\u001b[31m×\u001b[0m Encountered error while trying to install package.\n",
      "\u001b[31m╰─>\u001b[0m sklearn-contrib-py-earth\n",
      "\n",
      "\u001b[1;35mnote\u001b[0m: This is an issue with the package mentioned above, not pip.\n",
      "\u001b[1;36mhint\u001b[0m: See above for output from the failure.\n",
      "\u001b[?25h"
     ]
    }
   ],
   "source": [
    "!pip install sklearn-contrib-py-earth"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3 (H2O AI Cloud)",
   "language": "python",
   "name": "haic"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
